// @everymatrix/om-connector/src/tcp/tcp.utils.ts
import {
    ErrorResponse,
    InitialDataResponse,
    MessageResponse,
    PingRequest,
    SubscribeResponse,
    UnsubscribeResponse,
    UpdateDataResponse,
} from '../core/messages.types';
import isEntity from '../core/is-entity';

export function isPing(data: MessageResponse): data is PingRequest {
    return isEntity(data, 'PingRequest');
}

export function isSubscribeResponse(data: MessageResponse): data is SubscribeResponse {
    return isEntity(data, 'SubscribeResponse');
}

export function isInitialData(data: MessageResponse): data is InitialDataResponse {
    return isEntity(data, 'InitialData');
}

export function isUpdateData(data: MessageResponse): data is UpdateDataResponse {
    return isEntity(data, 'UpdateData');
}

export function isError(data: MessageResponse): data is ErrorResponse {
    return isEntity(data, 'error');
}

export function isUnsubscribeResponse(data: MessageResponse): data is UnsubscribeResponse {
    return isEntity(data, 'UnsubscribeResponse');
}
