// @everymatrix/om-connector/lib/constants/eventNames.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonEvents = exports.HTTPEvents = exports.TCPEvents = void 0;
var TCPEvents;
(function (TCPEvents) {
    TCPEvents["jsonReceived"] = "jsonReceivedTcp";
    TCPEvents["jsonUnzipped"] = "jsonUnzippedTcp";
    TCPEvents["notifyInitialDump"] = "notifyInitialDumpTcp";
    TCPEvents["notifyEntityUpdates"] = "notifyEntityUpdatesTcp";
    TCPEvents["ping"] = "pingTcp";
    TCPEvents["tcpError"] = "errorTcp";
    TCPEvents["subscribe"] = "subscribeTcp";
    TCPEvents["unsubscribe"] = "unsubscribeTcp";
})(TCPEvents = exports.TCPEvents || (exports.TCPEvents = {}));
var HTTPEvents;
(function (HTTPEvents) {
    HTTPEvents["jsonReceived"] = "jsonReceivedHttp";
    HTTPEvents["jsonUnzipped"] = "jsonUnzippedHttp";
    HTTPEvents["notifyInitialDump"] = "notifyInitialDumpHttp";
    HTTPEvents["notifyEntityUpdates"] = "notifyEntityUpdatesHttp";
    HTTPEvents["nextRequest"] = "makeNextRequestHttp";
    HTTPEvents["reconnect"] = "reconnectHttp";
    HTTPEvents["httpError"] = "httpError";
})(HTTPEvents = exports.HTTPEvents || (exports.HTTPEvents = {}));
var CommonEvents;
(function (CommonEvents) {
    CommonEvents["runtimeError"] = "runtimeError";
})(CommonEvents = exports.CommonEvents || (exports.CommonEvents = {}));
