// @everymatrix/om-connector/lib/http/http.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isError = exports.isUpdateData = exports.isInitialData = exports.isUnsubscribeResponse = exports.isSubscribeResponse = exports.HTTPFlowEvents = exports.RequestTypes = void 0;
var is_entity_1 = require("../core/is-entity");
var RequestTypes;
(function (RequestTypes) {
    RequestTypes["subscribe"] = "SubscribeRequest";
    RequestTypes["nextInitialData"] = "GetNextInitialDataRequest";
    RequestTypes["nextUpdateData"] = "GetNextUpdateDataRequest";
    RequestTypes["unsubscribe"] = "UnsubscribeRequest";
})(RequestTypes = exports.RequestTypes || (exports.RequestTypes = {}));
var HTTPFlowEvents;
(function (HTTPFlowEvents) {
    HTTPFlowEvents["subscriptionIdSet"] = "subscriptionIdSet";
    HTTPFlowEvents["dumpCompleted"] = "dumpCompleted";
})(HTTPFlowEvents = exports.HTTPFlowEvents || (exports.HTTPFlowEvents = {}));
function isSubscribeResponse(data) {
    return is_entity_1.default(data, 'SubscribeResponse');
}
exports.isSubscribeResponse = isSubscribeResponse;
function isUnsubscribeResponse(data) {
    return is_entity_1.default(data, 'UnsubscribeResponse');
}
exports.isUnsubscribeResponse = isUnsubscribeResponse;
function isInitialData(data) {
    return is_entity_1.default(data, 'GetNextInitialDataResponse');
}
exports.isInitialData = isInitialData;
function isUpdateData(data) {
    return is_entity_1.default(data, 'GetNextUpdateDataResponse');
}
exports.isUpdateData = isUpdateData;
function isError(data) {
    return is_entity_1.default(data, 'error');
}
exports.isError = isError;
