// @everymatrix/om-connector/lib/tcp/tcp-messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUpdateDataResumeRequest = exports.getUnsubscribeRequest = exports.getPingResponse = exports.getSubscriptionMessage = exports.getGzipPromise = void 0;
var zlib_1 = require("zlib");
var tcp_emitter_1 = require("./tcp-emitter/tcp-emitter");
var eventNames_1 = require("../constants/eventNames");
function getGzipPromise(content) {
    return new Promise(function (resolve, reject) {
        zlib_1.gzip(content, function (err, gzipped) {
            if (err) {
                tcp_emitter_1.tcpEmitter.emit(eventNames_1.CommonEvents.runtimeError, err);
                reject(err);
            }
            else {
                resolve(gzipped);
            }
        });
    });
}
exports.getGzipPromise = getGzipPromise;
function getMessage(requestMeta) {
    return getGzipPromise(JSON.stringify(requestMeta));
}
function getSubscriptionMessage(subscriptionName) {
    return getMessage({
        SubscribeRequest: {
            subscriptionSpecificationName: subscriptionName,
        },
    });
}
exports.getSubscriptionMessage = getSubscriptionMessage;
function getPingResponse(id) {
    return getMessage({
        PingResponse: {
            id: id,
        },
    });
}
exports.getPingResponse = getPingResponse;
function getUnsubscribeRequest(subscriptionId, subscriptionSpecificationName) {
    return getMessage({
        UnsubscribeRequest: {
            subscriptionId: subscriptionId,
            subscriptionSpecificationName: subscriptionSpecificationName,
        },
    });
}
exports.getUnsubscribeRequest = getUnsubscribeRequest;
function getUpdateDataResumeRequest(subscriptionId, subscriptionSpecificationName, subscriptionChecksum, lastBatchUuid) {
    return getMessage({
        UpdateDataResumeRequest: {
            subscriptionId: subscriptionId,
            subscriptionSpecificationName: subscriptionSpecificationName,
            subscriptionChecksum: subscriptionChecksum,
            lastBatchUuid: lastBatchUuid,
        },
    });
}
exports.getUpdateDataResumeRequest = getUpdateDataResumeRequest;
