// @everymatrix/om-connector/lib/tcp/tcp.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isUnsubscribeResponse = exports.isError = exports.isUpdateData = exports.isInitialData = exports.isSubscribeResponse = exports.isPing = void 0;
var is_entity_1 = require("../core/is-entity");
function isPing(data) {
    return is_entity_1.default(data, 'PingRequest');
}
exports.isPing = isPing;
function isSubscribeResponse(data) {
    return is_entity_1.default(data, 'SubscribeResponse');
}
exports.isSubscribeResponse = isSubscribeResponse;
function isInitialData(data) {
    return is_entity_1.default(data, 'InitialData');
}
exports.isInitialData = isInitialData;
function isUpdateData(data) {
    return is_entity_1.default(data, 'UpdateData');
}
exports.isUpdateData = isUpdateData;
function isError(data) {
    return is_entity_1.default(data, 'error');
}
exports.isError = isError;
function isUnsubscribeResponse(data) {
    return is_entity_1.default(data, 'UnsubscribeResponse');
}
exports.isUnsubscribeResponse = isUnsubscribeResponse;
